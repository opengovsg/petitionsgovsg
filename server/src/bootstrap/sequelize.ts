import { Sequelize } from 'sequelize'

import {
  defineSignature,
  definePost,
  defineAddressee,
  defineSubscription,
} from '@/models'
import { dbConfig } from './config/database'

export const sequelize = new Sequelize({ ...dbConfig, logging: false })

export const { Addressee } = defineAddressee(sequelize)
export const { Post } = definePost(sequelize, { Addressee })
export const Signature = defineSignature(sequelize, { Post })
export const Subscription = defineSubscription(sequelize, { Post })

export default sequelize
