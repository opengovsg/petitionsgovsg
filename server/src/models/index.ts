export { Signature, defineSignature } from './signatures.model'
export { Post, definePost } from './posts.model'
