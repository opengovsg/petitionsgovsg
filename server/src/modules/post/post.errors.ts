import { StatusCodes } from 'http-status-codes'
import { ApplicationError } from '../core/core.errors'

export class MissingPublicPostError extends ApplicationError {
  constructor(
    message = 'No public post with this id',
    statusCode = StatusCodes.NOT_FOUND,
  ) {
    super(message, statusCode)
  }
}

export class PostUpdateError extends ApplicationError {
  constructor(
    message = 'Post update failed',
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
  ) {
    super(message, statusCode)
  }
}
