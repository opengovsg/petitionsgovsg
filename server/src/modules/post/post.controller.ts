import { validationResult } from 'express-validator'
import { StatusCodes } from 'http-status-codes'
import { Post } from '~shared/types/base'
import { createLogger } from '../../bootstrap/logging'
import { Message } from '../../types/message-type'
import { ControllerHandler } from '../../types/response-handler'
import { SortType } from '../../types/sort-type'
import { AuthService } from '../auth/auth.service'
import { PostService, PostWithAddresseeAndSignatures } from './post.service'
import { hashData, generateSalt } from '../../util/hash'

const logger = createLogger(module)

export class PostController {
  private authService: Public<AuthService>
  private postService: Public<PostService>

  constructor({
    authService,
    postService,
  }: {
    authService: Public<AuthService>
    postService: Public<PostService>
  }) {
    this.authService = authService
    this.postService = postService
  }

  /**
   * Lists all post
   * @query sort Sort by popularity or recent
   * @query size Number of posts to return
   * @query page If size is given, specify which page to return
   * @return 200 with posts and totalItem for pagination
   * @return 500 when database error occurs
   */
  listPosts: ControllerHandler<
    undefined,
    { posts: Post[]; totalItems: number } | Message,
    undefined,
    {
      page?: number
      size?: number
      sort?: SortType
    }
  > = async (req, res) => {
    const { page, size, sort = SortType.Top } = req.query
    try {
      const data = await this.postService.listPosts({
        sort: sort as SortType,
        page: page,
        size: size,
      })
      return res.status(StatusCodes.OK).json(data)
    } catch (error) {
      if (error instanceof Error) {
        logger.error({
          message: 'Error while listing posts',
          meta: {
            function: 'listPosts',
          },
          error,
        })
        return res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Server Error' })
      }
    }
  }

  /**
   * Get a single post and all the users associated with it
   * @param postId Id of the post
   * @query relatedPosts if true, return related posts
   * @return 200 with post
   * @return 403 if user does not have permission to access post
   * @return 500 for database error
   */
  getSinglePost: ControllerHandler<
    { id: number },
    PostWithAddresseeAndSignatures | Message,
    undefined,
    { relatedPosts?: number }
  > = async (req, res) => {
    const errors = validationResult(req)
    if (!errors.isEmpty()) {
      return res.status(StatusCodes.BAD_REQUEST).json(errors.array()[0].msg)
    }
    let post
    try {
      post = await this.postService.getSinglePost(req.params.id)
    } catch (error) {
      logger.error({
        message: 'Error while retrieving single post',
        meta: {
          function: 'getSinglePost',
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server Error' })
    }

    try {
      await this.authService.verifyUserCanViewPost(
        post,
        req.user?.id.toString() ?? '',
      )
    } catch (error) {
      logger.error({
        message: 'Error while retrieving single post',
        meta: {
          function: 'getSinglePost',
        },
        error,
      })
      return res
        .status(StatusCodes.FORBIDDEN)
        .json({ message: 'User does not have permission to access this post' })
    }

    return res.status(StatusCodes.OK).json(post)
  }

  /**
   * Create a new post
   * @body title title of post
   * @body summary summary of post
   * @body reason reason of post
   * @body request request of post
   * @body references references of post
   * @return 200 if post is created
   * @return 400 if title and description is too short or long
   * @return 401 if user is not signed in
   * @return 500 if database error
   */
  createPost: ControllerHandler<
    undefined,
    { data: number } | Message,
    {
      title: string
      summary: string | null
      reason: string
      request: string
      references: string | null
      fullname: string
      addresseeId: number
      profile: string | null
      email: string
      salt: string
    },
    undefined
  > = async (req, res) => {
    const errors = validationResult(req)
    if (!errors.isEmpty()) {
      return res.status(StatusCodes.BAD_REQUEST).json(errors.array()[0].msg)
    }
    if (!req.user) {
      return res
        .status(StatusCodes.UNAUTHORIZED)
        .json({ message: 'User not signed in' })
    }

    try {
      if (!req.user) {
        return res
          .status(StatusCodes.UNAUTHORIZED)
          .json({ message: 'User not signed in' })
      }

      const salt = await generateSalt()
      const hashedUserSgid = await hashData(req.user.id, salt)
      const data = await this.postService.createPost({
        title: req.body.title,
        summary: req.body.summary,
        reason: req.body.reason,
        request: req.body.request,
        hashedUserSgid: hashedUserSgid,
        references: req.body.references,
        fullname: req.body.fullname,
        addresseeId: req.body.addresseeId,
        profile: req.body.profile,
        email: req.body.email,
        salt: salt,
      })

      return res.status(StatusCodes.OK).json({ data: data })
    } catch (error) {
      logger.error({
        message: 'Error while creating post',
        meta: {
          function: 'addPost',
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server error' })
    }
  }

  /**
   * Delete a post
   * @param id Post to be deleted
   * @return 200 if successful
   * @return 401 if user is not logged in
   * @return 403 if user does not have permission to delete post
   * @return 500 if database error
   */
  deletePost: ControllerHandler<{ id: string }, Message> = async (req, res) => {
    const postId = Number(req.params.id)
    try {
      const userId = req.user?.id
      if (!userId) {
        logger.error({
          message: 'UserId is undefined after authenticated',
          meta: {
            function: 'deletePost',
          },
        })
        return res
          .status(StatusCodes.UNAUTHORIZED)
          .json({ message: 'You must be logged in to delete posts.' })
      }
      // TODO
      // const hasPermission = await this.authService.hasPermissionToEditPost(
      //   userId,
      //   postId,
      // )
      // if (!hasPermission) {
      //   return res
      //     .status(StatusCodes.FORBIDDEN)
      //     .json({ message: 'You do not have permission to delete this post.' })
      // }
      await this.postService.deletePost(postId)
      return res.status(StatusCodes.OK).send({ message: 'OK' })
    } catch (error) {
      logger.error({
        message: 'Error while deleting post',
        meta: {
          function: 'deletePost',
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server error' })
    }
  }
}
