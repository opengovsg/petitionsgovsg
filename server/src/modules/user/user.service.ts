import { User } from '~shared/types/base'
import { ModelDef } from '../../types/sequelize'
import { LoadPublicUserDto } from '~shared/types/api'
export class UserService {
  private User: ModelDef<User>

  constructor({ User }: { User: ModelDef<User> }) {
    this.User = User
  }

  loadUser = async (userId: number): Promise<LoadPublicUserDto> => {
    return this.User.findByPk(userId) as Promise<LoadPublicUserDto>
  }

  loadFullUser = async (userId: number): Promise<User> => {
    return this.User.findByPk(userId) as Promise<User>
  }

  loadUserBySgid = async (sgid: string): Promise<LoadPublicUserDto> => {
    return this.User.findOne({ where: { sgid: sgid } })
  }
}
