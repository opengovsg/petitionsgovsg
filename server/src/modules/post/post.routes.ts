import { AuthMiddleware } from '../auth/auth.middleware'
import express from 'express'
import { check, param } from 'express-validator'
import { PostController } from './post.controller'

export const routePosts = ({
  controller,
  authMiddleware,
}: {
  controller: PostController
  authMiddleware: Public<AuthMiddleware>
}): express.Router => {
  const router = express.Router()

  /**
   * Lists all post
   * @route   GET /api/posts
   * @returns 200 with posts
   * @returns 500 when database error occurs
   * @access  Public
   */
  router.get('/', controller.listPosts)

  /**
   * Get a single post and all users associated with it
   * @route  GET /api/posts/:id
   * @return 200 with post
   * @return 403 if user does not have permission to access post
   * @return 500 for database error
   * @access Public
   */
  router.get('/:id', [param('id').isInt().toInt()], controller.getSinglePost)

  /**
   * Create a new post
   * @route  POST /api/posts/
   * @return 200 if post is created
   * @return 400 if title and description is too short or long
   * @return 401 if user is not signed in
   * @return 500 if database error
   * @access Private
   */
  router.post(
    '/',
    [
      authMiddleware.authenticate,
      check(
        'title',
        'Enter a title with minimum of 15 characters and maximum 150 characters',
      ).isLength({
        min: 15,
        max: 150,
      }),
      check('email', 'Enter an email with a minimum of 1 characters').isLength({
        min: 1,
      }),
    ],
    controller.createPost,
  )

  /**
   * Delete a post
   * @route  DELETE /api/posts/:id
   * @return 200 if successful
   * @return 401 if user is not logged in
   * @return 403 if user does not have permission to delete post
   * @return 500 if database error
   * @access Private
   */
  router.delete(
    '/:id([0-9]+$)',
    [authMiddleware.authenticate],
    controller.deletePost,
  )

  return router
}
