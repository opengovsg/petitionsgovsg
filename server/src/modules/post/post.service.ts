import type { Sequelize as SequelizeType } from 'sequelize'
import Sequelize, { Model, Op, OrderItem, ProjectionAlias } from 'sequelize'
import {
  ListPostsDto,
  PostWithAddresseeAndSignatures,
  UpdatePostReqDto,
} from '~shared/types/api'
import { Addressee, Post, PostStatus, Signature } from '~shared/types/base'
import { ModelDef } from '@/types/sequelize'
import { SortType } from '~shared/types/base/post'
import {
  AddresseeDoesNotExistError,
  MissingPublicPostError,
  PostUpdateError,
} from './post.errors'

export class PostService {
  private Signature: ModelDef<Signature>
  private Post: ModelDef<Post>
  private Addressee: ModelDef<Addressee>
  private sequelize: SequelizeType

  constructor({
    Signature,
    Post,
    Addressee,
    sequelize,
  }: {
    Signature: ModelDef<Signature>
    Post: ModelDef<Post>
    Addressee: ModelDef<Addressee>
    sequelize: SequelizeType
  }) {
    this.Signature = Signature
    this.Post = Post
    this.Addressee = Addressee
    this.sequelize = sequelize
  }

  private signatureCountLiteral: ProjectionAlias = [
    Sequelize.literal(`(
      SELECT COUNT(*)
      FROM signatures AS signature
      WHERE
        signature.postId = post.id
    )`),
    'signatureCount',
  ]

  private sortFunction = (sortType: SortType): OrderItem => {
    const sortDict = {
      [SortType.Newest]: ['createdAt', 'DESC'],
      [SortType.Oldest]: ['createdAt', 'ASC'],
      [SortType.MostSignatures]: [Sequelize.literal('signatureCount'), 'DESC'],
      [SortType.LeastSignatures]: [Sequelize.literal('signatureCount'), 'ASC'],
    }

    return sortDict[sortType] as OrderItem
  }
  /**
   * Return the paginated posts
   * @param posts post array to be paginated
   * @param page If size is given, specify which page to return
   * @param size Number of posts to return
   * @returns paginated post array and total items in the original array
   */
  private getPaginatedPosts = (
    posts: PostWithAddresseeAndSignatures[],
    page?: number,
    size?: number,
  ): {
    posts: PostWithAddresseeAndSignatures[]
    totalItems: number
  } => {
    const totalItems = posts.length
    let returnPosts = posts
    if (size) {
      const offset = page ? (page - 1) * size : 0
      returnPosts = returnPosts.slice(offset, Number(offset) + Number(size))
    }
    return {
      posts: returnPosts,
      totalItems: totalItems,
    }
  }

  /**
   * Lists all post
   * @param sort Sort by popularity or recent
   * @param size Number of posts to return
   * @param page If size is given, specify which page to return
   */
  listPosts = async ({
    sort,
    page,
    size,
  }: {
    sort: SortType
    page?: number
    size?: number
  }): Promise<ListPostsDto> => {
    const whereobj = {
      status: PostStatus.Open,
    }

    const orderarray = this.sortFunction(sort)

    const posts = (await this.Post.findAll({
      where: whereobj,
      order: [orderarray],
      include: [
        this.Signature,
        {
          model: this.Addressee,
          required: false,
          attributes: ['name', 'shortName'],
        },
      ],
      attributes: [
        'id',
        'createdAt',
        'updatedAt',
        'status',
        'title',
        'summary',
        'reason',
        'request',
        'hashedUserSgid',
        'references',
        'fullname',
        'addresseeId',
        'profile',
        'email',
        this.signatureCountLiteral,
      ],
    })) as (Model & PostWithAddresseeAndSignatures)[]

    if (!posts) {
      return { posts: [], totalItems: 0 }
    } else {
      return this.getPaginatedPosts(posts, page, size)
    }
  }

  /**
   * Get a single post and all users associated with it
   * @param postId Id of the post
   */
  getSinglePost = async (
    postId: number,
  ): Promise<PostWithAddresseeAndSignatures> => {
    const post = (await this.Post.findOne({
      where: {
        [Op.or]: [{ status: PostStatus.Open }, { status: PostStatus.Draft }],
        id: postId,
      },
      include: [
        this.Signature,
        {
          model: this.Addressee,
          required: false,
          attributes: ['name', 'shortName'],
        },
      ],
      attributes: [
        'createdAt',
        'updatedAt',
        'status',
        'title',
        'summary',
        'reason',
        'request',
        'hashedUserSgid',
        'references',
        'salt',
        'fullname',
        'addresseeId',
        'profile',
        'email',
        this.signatureCountLiteral,
      ],
    })) as Model & PostWithAddresseeAndSignatures
    if (!post) {
      throw new MissingPublicPostError()
    } else {
      return post
    }
  }

  /**
   * Create a new post
   * @param newPost Post to be created
   * @returns Id of the new post if it is successfully created
   */
  createPost = async (newPost: {
    title: string
    summary: string | null
    reason: string
    request: string
    hashedUserSgid: string
    references: string | null
    fullname: string
    addresseeId: number
    profile: string | null
    email: string
    salt: string
  }): Promise<number> => {
    // Check if addresseeId is valid
    const addresseeExists = await this.Addressee.findByPk(newPost.addresseeId)
    if (!addresseeExists) {
      throw new AddresseeDoesNotExistError()
    }
    const postId = await this.sequelize.transaction(async (transaction) => {
      const post = await this.Post.create(
        { ...newPost, status: PostStatus.Draft },
        { transaction },
      )
      return post.id
    })
    return postId
  }

  /**
   * Delete a post by setting the post to archive
   * @param id Post to be deleted
   * @returns void if successful
   */
  deletePost = async (id: number): Promise<void> => {
    try {
      await this.sequelize.transaction(async (transaction) => {
        const [dbUpdate] = await this.Post.update(
          { status: PostStatus.Closed },
          { where: { id: id }, transaction },
        )
        if (!dbUpdate) {
          throw new PostUpdateError()
        }
      })
    } catch (error) {
      throw error
    }
  }

  /**
   * Update a post
   * @param id Post to be updated
   * @returns boolean Indicate if update was successful
   */
  updatePost = async ({
    id,
    title,
    summary,
    reason,
    request,
    references,
    addresseeId,
    profile,
    email,
  }: UpdatePostReqDto & { id: number }): Promise<boolean> => {
    // Check if addresseeId is valid
    const addresseeExists = await this.Addressee.findByPk(addresseeId)
    if (!addresseeExists) {
      throw new AddresseeDoesNotExistError()
    }

    const updated = await this.sequelize.transaction(async (transaction) => {
      const dbUpdate = await this.Post.update(
        {
          title,
          summary,
          reason,
          request,
          references,
          addresseeId,
          profile,
          email,
        },
        {
          where: { id: id },
          transaction,
        },
      )
      return dbUpdate
    })
    return !!updated
  }

  /**
   * Change post status to Open
   * @param id Post to be published
   * @returns void if successful
   */
  publishPost = async (id: number): Promise<void> => {
    await this.sequelize.transaction(async (transaction) => {
      const [dbUpdate] = await this.Post.update(
        { status: PostStatus.Open },
        { where: { id: id }, transaction },
      )
      if (!dbUpdate) {
        throw new PostUpdateError()
      }
    })
  }
}
