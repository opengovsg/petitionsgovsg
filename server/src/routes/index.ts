import express from 'express'
import { OwnershipCheck } from '../middleware/checkOwnership'
import { SignatureController } from '../modules/signatures/signature.controller'
import { routeSignatures } from '../modules/signatures/signature.routes'
import { AuthController } from '../modules/auth/auth.controller'
import { AuthMiddleware } from '../modules/auth/auth.middleware'
import { routeAuth } from '../modules/auth/auth.routes'
import { PostController } from '../modules/post/post.controller'
import { routePosts } from '../modules/post/post.routes'
import { routeAddressees } from '../modules/addressee/addressee.routes'
import { AddresseeController } from '../modules/addressee/addressee.controller'

type ApiRouterOptions = {
  signature: {
    controller: SignatureController
    authMiddleware: AuthMiddleware
    checkOwnership: OwnershipCheck
  }
  auth: {
    controller: AuthController
    authMiddleware: AuthMiddleware
  }
  post: {
    controller: PostController
    authMiddleware: AuthMiddleware
  }
  addressee: {
    controller: AddresseeController
  }
}

export const api = (options: ApiRouterOptions): express.Router => {
  const router = express.Router()

  router.use('/auth', routeAuth(options.auth))
  router.use('/posts', routePosts(options.post))
  router.use('/posts/signatures', routeSignatures(options.signature))
  router.use('/addressees', routeAddressees(options.addressee))

  return router
}
