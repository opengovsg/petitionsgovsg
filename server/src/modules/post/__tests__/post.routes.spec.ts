import express from 'express'
import { StatusCodes } from 'http-status-codes'
import { Sequelize, ModelCtor } from 'sequelize'
import supertest from 'supertest'
import { Post, PostStatus } from '~shared/types/base'
import { UserAuthType } from '~shared/types/api'
import { Signature as SignatureModel, User as UserModel } from '../../../models'
import { ControllerHandler } from '../../../types/response-handler'
import { ModelDef } from '../../../types/sequelize'
import {
  createTestDatabase,
  getModel,
  getModelDef,
  ModelName,
} from '../../../util/jest-db'
import { UserService } from '../../user/user.service'
import { PostController } from '../post.controller'
import { routePosts } from '../post.routes'
import { PostService } from '../post.service'

describe('/posts', () => {
  let db: Sequelize
  let Signature: ModelCtor<SignatureModel>
  let Post: ModelDef<Post>
  let User: ModelCtor<UserModel>

  let userService: UserService
  let postService: PostService
  let controller: PostController

  const mockPosts: Post[] = []
  let mockUser: UserModel

  // Set up service, controller and route
  const authService = {
    hasPermissionToEditPost: jest.fn(),
    verifyUserCanViewPost: jest.fn(),
  }

  const authenticate: ControllerHandler = (req, res, next) => {
    req.user = { id: mockUser.id, type: UserAuthType.Public }
    next()
  }

  const authMiddleware = {
    authenticate,
  }

  beforeAll(async () => {
    db = await createTestDatabase()
    Signature = getModel<SignatureModel>(db, ModelName.Signature)
    Post = getModelDef<Post>(db, ModelName.Post)
    User = getModel<UserModel>(db, ModelName.User)
    userService = new UserService({ User })
    postService = new PostService({
      Signature,
      Post,
      User,
      sequelize: db,
    })
    mockUser = await User.create({
      sgid: 'u=34',
      email: 'tanahwee@test.gov.sg',
      fullname: 'Tan Ah Wee',
      active: true,
    })
    for (let title = 1; title <= 20; title++) {
      const mockPost = await Post.create({
        title: title.toString(),
        summary: 'summary',
        status: PostStatus.Open,
        userId: mockUser.id,
        reason: null,
        request: null,
        references: null,
        fullname: 'Mr. Tan',
      })
      mockPosts.push(mockPost)
    }
    controller = new PostController({ userService, authService, postService })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  afterAll(async () => {
    await db.close()
  })

  describe('GET /posts', () => {
    it('returns 200 and data on posts on request', async () => {
      const path = '/posts'
      const app = express()
      app.use(express.json())
      app.use(authenticate)
      app.use(path, routePosts({ controller, authMiddleware }))
      const request = supertest(app)

      // Act
      const response = await request.get(path)

      // Assert
      expect(response.status).toEqual(StatusCodes.OK)
      expect(response.body.posts.length).toStrictEqual(mockPosts.length)
      expect(response.body.totalItems).toStrictEqual(mockPosts.length)
    })
  })

  describe('POST /posts', () => {
    it('returns 200 and creates a post on request', async () => {
      const path = '/posts'
      const app = express()
      app.use(express.json())
      app.use(authenticate)
      app.use(path, routePosts({ controller, authMiddleware }))
      const request = supertest(app)

      const body = {
        title: 'A title of at least 15 characters',
        summary:
          'A summary of at least 50 characters is the minimum requirement.',
        fullname: 'Lim Yong Xiang',
      }

      // Act
      const {
        status,
        body: { data: postId },
      } = await request.post(path).send(body)
      const post = await Post.findByPk(postId)

      // Assert
      expect(status).toEqual(StatusCodes.OK)
      expect(post).toBeDefined()
    })

    it('returns 403 if the title and description do not fulfil the character requirements', async () => {
      const path = '/posts'
      const app = express()
      app.use(express.json())
      app.use(authenticate)
      app.use(path, routePosts({ controller, authMiddleware }))
      const request = supertest(app)

      const body = {
        title: 'badTitle',
        summary: 'badSummary',
      }

      const response = await request.post(path).send(body)

      // Assert
      expect(response.status).toEqual(StatusCodes.BAD_REQUEST)
    })
  })

  describe('DELETE /posts/:id', () => {
    it('returns 200 if post is deleted', async () => {
      const { id } = mockPosts[0]
      authService.hasPermissionToEditPost.mockResolvedValue(true)

      const path = '/posts'
      const app = express()
      app.use(express.json())
      app.use(authenticate)
      app.use(path, routePosts({ controller, authMiddleware }))
      const request = supertest(app)

      const response = await request.delete(path + `/${id}`)

      expect(response.status).toEqual(StatusCodes.OK)
    })
  })
})
