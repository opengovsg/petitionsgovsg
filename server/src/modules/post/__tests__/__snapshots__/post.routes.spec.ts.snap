// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/posts Get / retrieves all posts 1`] = `
Object {
  "posts": Array [
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": "2022-01-01T00:00:00.000Z",
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
      "id": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
      "profile": null,
      "reason": "reason for",
      "references": null,
      "request": "request",
      "signatureCount": 3,
      "signatures": Array [
        Object {
          "comment": null,
          "createdAt": "2022-01-01T00:00:00.000Z",
          "fullname": null,
          "hashedUserSgid": "1234",
          "id": 1,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": "2022-01-01T00:00:00.000Z",
        },
        Object {
          "comment": null,
          "createdAt": "2022-01-01T00:00:00.000Z",
          "fullname": "John Doe",
          "hashedUserSgid": "5678",
          "id": 2,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": "2022-01-01T00:00:00.000Z",
        },
        Object {
          "comment": "I support this cause!",
          "createdAt": "2022-01-01T00:00:00.000Z",
          "fullname": null,
          "hashedUserSgid": "8765",
          "id": 3,
          "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
          "type": "support",
          "updatedAt": "2022-01-01T00:00:00.000Z",
        },
      ],
      "status": "OPEN",
      "summary": null,
      "title": "Basic petition without optional items",
      "updatedAt": "2022-01-01T00:00:00.000Z",
    },
    Object {
      "addressee": Object {
        "name": "Ministry of Environment",
        "shortName": "MoE",
      },
      "addresseeId": 1,
      "createdAt": "2022-01-01T00:00:00.000Z",
      "email": "John@email.com",
      "fullname": "John Doe",
      "hashedUserSgid": "wxyz",
      "id": "aaaaaaaa-dea2-4f78-8dc4-5542cafaff87",
      "profile": "OGP",
      "reason": "reason",
      "references": "wikipedia.com",
      "request": "request",
      "signatureCount": 0,
      "signatures": Array [],
      "status": "OPEN",
      "summary": "petition summary",
      "title": "Petition with summary, references, profile",
      "updatedAt": "2022-01-01T00:00:00.000Z",
    },
  ],
  "totalItems": 2,
}
`;

exports[`/posts Get /:id retrieves posts with specified id 1`] = `
Object {
  "addressee": Object {
    "name": "Ministry of Environment",
    "shortName": "MoE",
  },
  "addresseeId": 1,
  "createdAt": "2022-01-01T00:00:00.000Z",
  "email": "John@email.com",
  "fullname": "John Doe",
  "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
  "profile": null,
  "reason": "reason for",
  "references": null,
  "request": "request",
  "salt": "$2b$10$72TKINiR4.ZMaJZmuN5tW.",
  "signatureCount": 3,
  "signatures": Array [
    Object {
      "comment": null,
      "createdAt": "2022-01-01T00:00:00.000Z",
      "fullname": null,
      "hashedUserSgid": "1234",
      "id": 1,
      "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
      "type": "support",
      "updatedAt": "2022-01-01T00:00:00.000Z",
    },
    Object {
      "comment": null,
      "createdAt": "2022-01-01T00:00:00.000Z",
      "fullname": "John Doe",
      "hashedUserSgid": "5678",
      "id": 2,
      "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
      "type": "support",
      "updatedAt": "2022-01-01T00:00:00.000Z",
    },
    Object {
      "comment": "I support this cause!",
      "createdAt": "2022-01-01T00:00:00.000Z",
      "fullname": null,
      "hashedUserSgid": "8765",
      "id": 3,
      "postId": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
      "type": "support",
      "updatedAt": "2022-01-01T00:00:00.000Z",
    },
  ],
  "status": "OPEN",
  "summary": null,
  "title": "Basic petition without optional items",
  "updatedAt": "2022-01-01T00:00:00.000Z",
}
`;

exports[`/posts Post / returns BAD REQUEST on bad request 1`] = `"Enter a title with minimum of 15 characters and maximum 150 characters"`;

exports[`/posts Post / returns OK on valid submission 1`] = `
Object {
  "data": Any<String>,
}
`;

exports[`/posts Post / returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in",
}
`;

exports[`/posts Put /:id returns BAD REQUEST on bad request 1`] = `"Enter a title with minimum of 15 characters and maximum 150 characters"`;

exports[`/posts Put /:id returns OK on valid submission 1`] = `
Object {
  "message": "Petition updated",
}
`;

exports[`/posts Put /:id returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "You must be logged in to update posts.",
}
`;
